Testing: Ingest expected files and input workfile

tokejepsen opened the pull request:
## Changelog Description
This ingests the Maya workfile from the Drive storage. Have changed the format to MayaAscii so its easier to see what changes are happening in a PR. This meant changing the expected files and database entries as well.

## Testing notes:
```
.\openpype_console.bat runtests C:\Users\tokejepsen\OpenPype\tests\integration\hosts\maya\test_publish_in_maya.py --mongo_url "mongodb://localhost:2707/"
```

Split from https://github.com/ynput/OpenPype/pull/5644

kalisp commented:
Implemented LocalFileHandler to load from folder in repo. (And to match all other folder comparison.)
Test name added to structure to separate files for each test.
<img width="960" alt="AnyDesk_Y8HKgi065P" src="https://github.com/ynput/OpenPype/assets/4457962/89983bc1-bd2b-4386-94a3-d46945558c5d">

Removed superfluous `files` to keep other Maya test working

tokejepsen commented:
@kalisp nice solution!

I've fixed some issues and added a suggestion to skip directories when checking the folder structure, otherwise I get this error when testing Maya:
```
___________________________ TestPublishInMaya.test_folder_structure_same ____________________________

self = <test_publish_in_maya.TestPublishInMaya object at 0x000002E6D52E2760>
dbcon = <openpype.pipeline.mongodb.AvalonMongoDB object at 0x000002E6D50BBF40>
publish_finished = True, download_test_data = 'C:\\Users\\TOKEJE~1\\AppData\\Local\\Temp\\tmpc71s9xn6'
output_folder_url = 'C:\\Users\\TOKEJE~1\\AppData\\Local\\Temp\\tmpc71s9xn6\\output'
skip_compare_folders = [], setup_only = None

    def test_folder_structure_same(self, dbcon, publish_finished,
                                   download_test_data, output_folder_url,
                                   skip_compare_folders,
                                   setup_only):
        """Check if expected and published subfolders contain same files.

            Compares only presence, not size nor content!
        """
        if setup_only or self.SETUP_ONLY:
            print("Creating only setup for test, not launching app")
            return

        published_dir_base = output_folder_url
        expected_dir_base = os.path.join(download_test_data,
                                         "expected")

        print(
            "Comparing published:'{}' : expected:'{}'".format(
                published_dir_base, expected_dir_base
            )
        )

        def get_files(dir_base):
            result = set()

            for f in glob.glob(dir_base + "\\**", recursive=True):
                #if os.path.isdir(f):
                #    continue

                if f != dir_base and os.path.exists(f):
                    result.add(f.replace(dir_base, ""))

            return result

        published = get_files(published_dir_base)
        expected = get_files(expected_dir_base)

        filtered_published = self._filter_files(
            published, skip_compare_folders
        )

        # filter out temp files also in expected
        # could be polluted by accident by copying 'output' to zip file
        filtered_expected = self._filter_files(expected, skip_compare_folders)

        not_matched = filtered_expected.symmetric_difference(
            filtered_published
        )
        if not_matched:
>           raise AssertionError(
                "Missing {} files".format("\n".join(sorted(not_matched)))
            )
E           AssertionError: Missing \test_project\test_asset\work\test_task\cache
E           \test_project\test_asset\work\test_task\cache\alembic files

..\tests\lib\testing_classes.py:413: AssertionError
```
We could maybe look at preventing Maya from creating the empty folder in the first place, but I dont think that is worth the effort. And is there any point in testing for empty folders?

kalisp commented:
I think there is no need to test empty folders, just wondering why it haven't happened for me.

BigRoy commented on tests/lib/testing_classes.py:
```diff
@@ -365,26 +377,42 @@ def test_folder_structure_same(self, dbcon, publish_finished,
-        if not_mtched:
-            raise AssertionError("Missing {} files".format(
-                "\n".join(sorted(not_mtched))))
+        not_matched = filtered_expected.symmetric_difference(
+            filtered_published
+        )
+        if not_matched:
+            raise AssertionError(
+                "Missing {} files".format("\n".join(sorted(not_matched)))
+            )
```
This error message is incorrect - right? It's not necessarily missing files but would also error if it generated more files than expected.

Likely we're best off reporting it as:

- Generated unexpected files: {unexpected}
- Missing expected files: {missing}

Just so the report is clear as to what's going on.

kalisp replied:
https://github.com/ynput/OpenPype/issues/5889

BigRoy commented on tests/lib/testing_classes.py:
```diff
@@ -365,26 +377,42 @@ def test_folder_structure_same(self, dbcon, publish_finished,
+        def get_files(dir_base):
+            result = set()
+
+            for f in glob.glob(dir_base + "\\**", recursive=True):
```
Since you are basically globbing all recursively why not use `os.walk`? (or `os.scandir` is that's still a thing that it's faster)

tokejepsen replied:
Is using `os.walk` for speed?

BigRoy replied:
I thought it might also improve readability - but I also suspect it's faster as it doesn't need to do the string comparisons, etc + you don't need the additional `isdir` check.

Speed-wise I don't think you'll really notice anything for the amount of files we're likely processing here.

kalisp replied:
Would be interested in how big performance gains it would produce. It looks more like fixing plastic pollution by banning of plastic straws ;).
But feel free to modify it, irrelevant to this PR though.

BigRoy replied:
Agree - was just wondering initially why `glob` was used.

kalisp replied:
Not sure, I think I was getting frustrated by seemingly simple use case that I expected would have been already solved zillion times by some built-in or third party lib.

BigRoy commented on tests/lib/testing_classes.py:
```diff
@@ -365,26 +377,42 @@ def test_folder_structure_same(self, dbcon, publish_finished,
-                       glob.glob(expected_dir_base + "\\**", recursive=True)
-                       if f != expected_dir_base and os.path.exists(f))
+        print(
+            "Comparing published:'{}' : expected:'{}'".format(
```
```suggestion
            "Comparing published: '{}' | expected: '{}'".format(
```

BigRoy commented on tests/lib/testing_classes.py:
```diff
@@ -80,14 +81,25 @@ def download_test_data(self, test_data_folder, persist, request):
                 file_id, file_name, md5 = test_file
 
-                f_name, ext = os.path.splitext(file_name)
+                CURR_DIR = os.path.dirname(os.path.abspath(
+                    inspect.getfile(self.__class__)))
+                if os.path.exists(file_id):
+                    handler_class = LocalFileHandler
+                elif os.path.exists(os.path.join(CURR_DIR, file_id)):
+                    file_id = os.path.join(CURR_DIR, file_id)
```
Why is `CURR_DIR` all caps?

tokejepsen replied:
@kalisp

BigRoy commented on tests/integration/hosts/maya/test_publish_in_maya/input/dumps/openpype_tests/logs.bson:
Do we really need the logs to kickstart the tests?

BigRoy commented on tests/integration/hosts/maya/test_publish_in_maya/input/dumps/avalon_tests/test_project.bson:
Are there docs available on how to generate dumps like these? 

Should each dump also include a `readme` as to WHAT it generates as data for someone writing tests against it?

tokejepsen replied:
> Are there docs available on how to generate dumps like these?

@kalisp generated these.
It's on my list (https://github.com/ynput/OpenPype/pull/5644) to have a command flag for dumping the database.

> Should each dump also include a readme as to WHAT it generates as data for someone writing tests against it?

As part of the my task above to easier dump the database, I'll be looking into dumping to a human readable json, similar to https://github.com/ynput/OpenPype/blob/15e82c8af836283561ee975fd7bc1741fbc1b691/tests/integration/hosts/maya/expected/avalon_tests.test_project.json

kalisp replied:
![image](https://github.com/ynput/OpenPype/assets/4457962/ba7a5b5f-af26-4065-95e4-316f05e3da64)
(From `tests/integration/README.md`)

BigRoy commented on tests/integration/hosts/maya/test_publish_in_maya/input/workfile/test_project_test_asset_test_task_v001.ma:
```diff
@@ -0,0 +1,476 @@
+fileInfo "version" "2022";
+fileInfo "cutIdentifier" "202205171752-c25c06f306";
+fileInfo "osv" "Windows 10 Pro v2009 (Build: 19044)";
+fileInfo "license" "education";
```
Won't educational licensed files start adding all kinds of license warnings when running tests?

tokejepsen replied:
Didnt get any issues running the tests here.

kalisp commented:
Does this still depend on https://github.com/ynput/OpenPype/pull/5785 ? I think that one should be closed and only this PR used.

tokejepsen commented:
> Does this still depend on #5785 ? I think that one should be closed and only this PR used.

No its does not depend on #5785 anymore. Will close that PR when this gets merged.
